packages/pubsub/test/connection-pool.js

```diff
@@ -674,15 +750,162 @@ describe('ConnectionPool', function() {
     });
   });
 
+  describe('getAndEmitChannelState', function() {
+    var channelErrorEvent = 'channel.error';
+    var channelReadyEvent = 'channel.ready';
+    var channelReadyState = 2;
+    var fakeChannelState;
+    var dateNow;
+    var fakeTimestamp;
+
+    var fakeClient = {
+      getChannel: function() {
+        return {
+          getConnectivityState: function(shouldConnect) {
+            assert.strictEqual(shouldConnect, false);
+            return fakeChannelState;
+          }
+        };
+      }
+    };
+
+    before(function() {
+      dateNow = global.Date.now;
+    });
+
+    beforeEach(function() {
+      fakeChannelState = 0;
+      fakeClient.waitForReady = fakeUtil.noop;
+
+      pool.getClient = function(callback) {
+        callback(null, fakeClient);
+      };
+
+      fakeTimestamp = dateNow.call(global.Date);
+      pool.noConnectionsTime = 0;
+
+      global.Date.now = function() {
+        return fakeTimestamp;
+      };
+    });
+
+    after(function() {
+      global.Date.now = dateNow;
+    });
+
+    it('should set the isGettingChannelState flag to true', function() {
+      pool.getAndEmitChannelState();
+      assert.strictEqual(pool.isGettingChannelState, true);
+    });
+
+    it('should emit any client errors', function(done) {
+      var channelErrorEmitted = false;
+
+      pool.on(channelErrorEvent, function() {
+        channelErrorEmitted = true;
+      });
+
+      var fakeError = new Error('nope');
+      var errorEmitted = false;
+
+      pool.on('error', function(err) {
+        assert.strictEqual(err, fakeError);
+        errorEmitted = true;
+      });
+
+      pool.getClient = function(callback) {
+        callback(fakeError);
+
+        assert.strictEqual(pool.isGettingChannelState, false);
```

Comment:
I'm somewhat confused why this batch of asserts isn't a victim of the event queue. `callback` sends out async events, which apparently are being executed synchronously before these asserts 😕 . Just to be safe, can we do a `setImmediate` wrapper around these 3 asserts and done()? Unless I'm missing something about why this works.

Reply:
Looking at https://github.com/nodejs/node/blob/master/lib/events.js#L207
and https://github.com/nodejs/node/blob/master/lib/events.js#L103 it seems as though they are not processed asynchronously.

Reply:
Fascinating! Never would have expected that.